### PR TITLE
TTD -- Move debugger explicit previous statement tracking into seperate interpreter loop implementation and optimize TTD flag checking.

### PR DESCRIPTION
**Pull request by mrkmarron**

### CODE REVIEW

**Comment by mrkmarron** [comment]
@dotnet-bot test Ubuntu ubuntu_linux_release please

**Comment by mrkmarron** [comment]
@agarwal-sandeep please take a look.

File: lib/Runtime/Base/ScriptContext.cpp [hunk 1]
```diff
@@ -140,6 +140,7 @@ namespace Js
 #if ENABLE_TTD
         , TTDHostCallbackFunctor()
         , TTDMode(TTD::TTDMode::Invalid)
+        , TTDShouldPerformRecordAction(false), TTDShouldPerformDebugAction(false), TTDShouldPerformRecordOrDebugAction(false)
```
Comment (agarwal-sandeep) [inline lib/Runtime/Base/ScriptContext.cpp]: >TTDShouldPerformDebugAction(false), TTDShouldPerformRecordOrDebugAction(false) [](start = 47, length = 78)

nit: in different lines to maintain code consistency

File: lib/Runtime/Base/ScriptContext.h [hunk 2]
```diff
@@ -1086,15 +1086,18 @@ namespace Js
 
         ////
 
-        //Use this to check specifically if we are in record AND this code is being run on behalf of the user application
-        bool ShouldPerformRecordAction() const
+        //Memoized results for frequently used Record/Replay action checks
+        bool TTDShouldPerformRecordOrDebugAction;
+        bool TTDShouldPerformRecordAction;
+        bool TTDShouldPerformDebugAction;
```
Comment (agarwal-sandeep) [inline lib/Runtime/Base/ScriptContext.h]: Can be bit fields right?

File: lib/Runtime/Language/InterpreterStackFrame.cpp [hunk 3]
```diff
@@ -1118,33 +1125,75 @@ namespace Js
 #if ENABLE_NATIVE_CODEGEN
             if (doJITLoopBody)
             {
-                newInstance->opProfiledLoopBodyStart = &InterpreterStackFrame::ProfiledLoopBodyStart<true, true>;
-                newInstance->opLoopBodyStart = &InterpreterStackFrame::LoopBodyStart<true, true>;
+                //right now we disallow stmt tracking fot TTD if JIT is enabled
```
Comment (agarwal-sandeep) [inline lib/Runtime/Language/InterpreterStackFrame.cpp]: >fot [](start = 54, length = 3)

nit: for

**Comment by digitalinfinity** [comment]
@mrkmarron what's the binary size impact of this change on ChakraCore?

**Comment by mrkmarron** [comment]
@digitalinfinity tl; dr; the size increase between No TTD and TTD Enabled in current master is 285kb, the inclusion of the new interpreter loop is an additional 55kb.

The sizes of binaries with various TTD code included are as follows:
Master with TTD disabled:    6,702kb
Master with Record only:      6,928kb
Master with Record Replay:  6,987kb
This PR with extra loop:        7,043kb

**Comment by pleath** [comment]
LGTM. Thanks.

**Comment by mrkmarron** [comment]
Subsumed by PR #1810 -- Merge of TTD Loop, Multi-Context, JIT support, and TTD check simplification.

**Pull request closed without merging**